Step2 - 연관 관계 매핑

lewisseo91 opened the pull request:
안녕하세요~!
리뷰 해 주셔서 감사합니다. 
연관 관계에 대해서 생각 해보고 진행하면서 오류도 엄청나고 하느라 PR이 조금 늦었네요 ㅎㅎ
아마 하고나서도 부족한 부분이 많겠다 생각하고 있지만 step2의 요구조건과 요구조건의 테스트는 완료 한 것 같아서 
리뷰 받고 싶어서 PR 올리게 되었습니다~!
감사합니다~!!

lewisseo91 committed:
fix(application): EnableJpaAuditing 설정

기존 테스트에 있던 EnableJpaAuditing 제거
Application EnableJpaAuditing 설정

lewisseo91 committed:
style(env): 내부 h2 테스트 전환

내부 h2 테스트 전환 (빠른 스키마 전환을 위해 테스트 할 때 동안만 시도)

lewisseo91 committed:
feat(docs): 현재 설계 시 고민 중인 부분 추가

현재 설계 시 고민 중인 부분 추가

lewisseo91 committed:
feat(docs): 현재 고민 부분 추가

현재 고민 부분 추가

lewisseo91 committed:
refactor(domain): 객체 연관 관계 매핑 추가

Answer
Long writerId -> User writer (일대일 단방향)
Long questionId -> Question question (다대일 양방향)
DeleteHistory
Long deletedById -> User deletedBy
QnaService
getWriterId -> getWriter
Question
Long writerId -> User writer (일대일 단방향)
answers 추가 (일대다 양방향 mappedBy)

lewisseo91 committed:
refactor(test): 객체 연관 관계 매핑 테스트 추가

global
static answer, static question, static deleteHistory 제거
import EnableJpaAuditing 패키지 제거
AnswerTest
answer save with question 추가 (연관 관계 매핑 테스트)
answer remove with question 삭제 (연관 관계 매핑 테스트)
DeleteHistoryTest
User 영속성 컨텍스트로 저장
QuestionTest
question save with answer 추가 (연관 관계 매핑 테스트)
question remove with answer 삭제 (연관 관계 매핑 테스트)
QnaServiceTest
서비스 테스트 중에 필요한 영속성 컨텍스트 변경
getWriterId -> getWriter
QuestionTest.Q1 -> question

lewisseo91 committed:
style(env): h2 외부 db 연동으로 변경

h2 외부 db 연동으로 변경 (내부에서 테스트 완료해서 변경 후 테스트 완료)

lewisseo91 committed:
style(question): toQuestion 네이밍 -> setQuestion 으로 변경

toQuestion 네이밍 -> setQuestion 으로 변경

lewisseo91 committed:
fix(AnswerTest): question에 writeBy 추가

userId 가 계속 null 이었음

lewisseo91 committed:
feat(docs): 관계 예상 그림 추가

3차 관계 예상 그림 추가

lewisseo91 committed:
chore(docs): 줄 바꿈 추가

줄 바꿈 추가

lewisseo91 committed:
chore(docs): 그림 관계 수정

그림 관계 수정

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -19,9 +22,13 @@
     private Long id;
 
-    private Long writerId;
+    @OneToOne
+    @JoinColumn(name = "writer_id")
+    private User writer;
```
`@OneToOne`으로 연관관계를 맺어주신 이유가 있으신가요?
`User` 한명이 여러개의 `Answer`을 달 수 있을 것 같은데요.

lewisseo91 replied:
맞습니다 
answer를 기준으로만 너무 생각해서 OneToOne 이라고 잘못 생각했네요 감사합니다~!

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -19,9 +22,13 @@
+    private User writer;
 
-    private Long questionId;
+    @ManyToOne
+    @JoinColumn(name = "question_id")
+    private Question question;
```
`Answer`를 조회시 항상 `Question, Writer`를 같이 조회해야 하는 경우가 아니라면 `지연로딩` 속성을 주면 좋겠네요.
`@OneToMany`는 LAZY가 디폴트이지만 `@ManyToOne`는 아니기 때문입니다.

lewisseo91 replied:
Lazy 꼭 걸겠습니다 감사합니다~!

slamdunk7575 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -21,11 +26,16 @@
     private String contents;
 
-    private Long writerId;
+    @OneToOne
+    @JoinColumn(name = "writer_id")
+    private User writer;
```
위의 피드백 https://github.com/next-step/jwp-qna/pull/294#discussion_r749145873 을 참고해주세요.

lewisseo91 replied:
감사합니다~!

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -47,29 +48,39 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-        this.writerId = writer.getId();
-        this.questionId = question.getId();
+        this.writer = writer;
+        this.question = question;
         this.contents = contents;
     }
```
`question의 answers` 에도 `answer` 를 추가해줘야 할 것 같은데요.
양방향 연관관계는 양쪽 객체 모두 신경써야하는 부분이 있습니다.

lewisseo91 replied:
감사합니다 추가 하도록 하겠습니다~ answer 의 setQuestion 쪽에서 add를 통해 추가하고 있었는데 이 부분은 answers 쪽에서 추가하도록 변경하겠습니다~!

아하 그리고 제가 잘못 추가하고 있었네요~! 감사합니다!

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -47,29 +48,39 @@ public Answer(Long id, User writer, Question question, String contents) {
     }
 
     public boolean isOwner(User writer) {
-        return this.writerId.equals(writer.getId());
+        return this.writer.equals(writer);
+    }
```
객체의 동등성 비교를 위해 `equals()`를 재정의해줘야 
영속성 컨텍스트 밖에서도 잘 동작할 것 같네요.

slamdunk7575 reviewed (changes requested):
연관관계 맵핑 잘해주셨네요.
조금 고민해보면 좋을 부분에 피드백 남겨드렸어요.
확인해보시고 리뷰 요청 부탁드립니다. 끝까지 화이팅 입니다! 💪

lewisseo91 committed:
feat(docs): 고민 추가

DeleteHistory를 몰라도 될까? 에 대한 고민 -> 유저가 Question, Answer, DeleteHistory를 몰라도 될까? 에 대한 고민

lewisseo91 committed:
fix(domain): 연관 관계 변경, FetchType LAZY 적용

연관 관계에 FetchType LAZY 적용
ManyToOne은 기본이 LAZY이지만 일단 발라 줌
question -> user
단방향 oneToOne (잘못 생각함) user가 여러개의 question을 필요로 할 수 있음
=> 단방향 manyToOne 로 변경 (양방향 고민 중)
answer -> user
단방향 oneToOne (잘못 생각함) user가 여러개의 answer를 필요로 할 수 있음
=> 단방향 manyToOne 로 변경 (양방향 고민 중)

lewisseo91 committed:
fix(domain): 연관 관계 메서드 추가 question <-> answer

question 쪽에만 적용되어있는 로직처럼 보여서 변경
question answers add 추가
answer changeQuestion setQuestion을 분리하여 적용

lewisseo91 committed:
feat(domain): 각 객체들 equals hashcode 추가

각 객체들 equals hashcode 추가

lewisseo91 committed:
style(AnswerTest): 네이밍 변경 적용

네이밍 변경 적용

lewisseo91 committed:
feat(user): user 쪽 question 양방향 연결

기존 question -> user
ManyToOne 단뱡향 => ManyToOne, OneToMany 양방향으로 변경

lewisseo91 committed:
fix(domain): answer, question null 체크 기준 변경

기존 question, writer -> this.question, this.writer 로 변경

lewisseo91 committed:
feat(UserTest): user quesetion add 테스트 추가

user question add 테스트 추가
update 쿼리 발생 확인

lewisseo91 committed:
feat(docs): 관계 예상 그림 4차 추가

관계 예상 그림 4차 추가 (1,2차 삭제)

lewisseo91 commented:
안녕하세요~! 
리뷰 해 주셔서 정말 감사합니다!!
리뷰어님 덕분에 매일 행복한 나날들을 보내고 있습니다.
말씀 해 주신 부분들 반영했습니다. 
하면서 고민했던 부분들 궁금해서 질문 남겼습니다~!

**궁금했던 점**

---

**1번**

제가 ManyToOne 관계를 Answer, Question, DeleteHistory 단방향을 설정하고 나서 user에서 '각각에 대해 양방향을 설정해야 할까?' 라는 부분이 굉장히 고민이었습니다. 
생각 해 본 결과 저는 아직 연관 관계에 대해서 어떻게 양방향, 단방향을 결정지을 지 잘 모른다고 생각하게 되었습니다.

a 와 b 연관 관계를 파악할 때 현재는

```
b는 1개의 a만 필요로 한다. OneToOne
a도 1개의 b만 필요로 한다. 
```

```
b는 여러 개의 a를 필요로 한다. ManyToOne
a는 1개의 b를 필요로 한다. OneToMany 양방향
a도 여러 개의 b를 필요로 한다. ManyToMany
```

이런 형태로 생각을 하면서 설계를 하고 있는데요. 
a는 ~개의 b를 필요로 한다. 라는 형태로 생각을 해서 연관 관계를 결정해도 괜찮을까요?
리뷰어님이 보통 연관 관계를 설정하실 때는 어떤 기준으로 하시는 지 궁금합니다!

**2번**

제가 user에서 Answer, Question, DeleteHistory 중 어떤 것을 양방향으로 설정해야 할 지 고민하다가 reddit 모방하자고 생각이 떠올라 사이트를 참고하게 되었습니다.

![reddit_exam](https://user-images.githubusercontent.com/17772475/141792117-9ee7bf53-cd98-40bd-ad2c-eb671cb2331c.PNG)

사이트에서 user는 question List를 요구하고 answer는 question에 귀속되게 나오는 것 같아서 question만 양방향으로 설정했는데 괜찮은 양방향 설정이었는지 궁금합니다~!

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -47,29 +49,43 @@ public Answer(Long id, User writer, Question question, String contents) {
-        this.writerId = writer.getId();
-        this.questionId = question.getId();
+        this.writer = writer;
+        this.question = question;
         this.contents = contents;
     }
```
이전 피드백 https://github.com/next-step/jwp-qna/pull/294#discussion_r749149055 을 드렸던 의도는 `답변(Answer)`이 생성될때 
`질문(Question)` 쪽에도 answers를 알고 있기 때문에 생성자에서 아래처럼 추가해보면 좋을것 같아 드렸던 거였어요.

`question.addAnswer(this);`

lewisseo91 replied:
아! 그렇네요 와 감사합니다!!

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -88,12 +104,29 @@ public void setDeleted(boolean deleted) {
+
     @Override
     public String toString() {
         return "Answer{" +
                 "id=" + id +
-                ", writerId=" + writerId +
-                ", questionId=" + questionId +
+                ", writerId=" + writer.getId() +
+                ", question=" + question.getId() +
                 ", contents='" + contents + '\'' +
                 ", deleted=" + deleted +
                 '}';
```
참고로 `toString()`메소드로 객체를 출력할때 양방향 연관 관계일 경우 
`순환 참조`가 일어날 수 있기 때문에 주의해야 합니다.

lewisseo91 replied:
감사합니다~! 유의하겠습니다

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -88,12 +104,29 @@ public void setDeleted(boolean deleted) {
         this.deleted = deleted;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Answer answer = (Answer) o;
+        return deleted == answer.deleted &&
+                Objects.equals(id, answer.id) &&
+                Objects.equals(writer, answer.writer) &&
+                Objects.equals(question, answer.question) &&
+                Objects.equals(contents, answer.contents);
+    }
```
조금 고민해볼 수 있는 부분이 `Entity` 클래스의 `equals()` 는 `id 값`으로 비교하는게 좋을 것 같은데요. 
보통 객체의 동등성 비교는 두 객체의 레퍼런스가 달라도 같은지 비교하는 것인데 
Entity 에서는 다른 필드는 equals() 비교에서 제외해도 된다고 생각합니다.

예를 들어, 1차 캐시를 초기화한 후에 다시 데이터베이스에서 동일한 엔티티를 읽어오는 경우 새로운 객체가 생성 되어도 두 객체를 equals() 했을 때 같아야 같은 row의 데이터임을 확인할 수 있기 때문입니다. 

하지만 일반적인 equals()와 다르게 id 값으로만 비교를 하고 아직 영속화되지 않은 Entity에 대한 고민도 해봐야 합니다. 
정답이 있는 것이 아니니 참고만 해보시면 좋을 것 같아요~

lewisseo91 replied:
엔티티에 대한 이해도가 많이 낮았네요 ㅠ
좋은 조언 감사합니다~! 
내용 정말 좋아서 메모하고 꼭 다시 볼만한 좋은 조언이네요!!
감사합니다~!!

slamdunk7575 commented on src/main/java/qna/domain/User.java:
```diff
@@ -31,6 +35,9 @@
     @Column(length = 50)
     private String email;
 
+    @OneToMany(mappedBy = "writer", fetch = FetchType.LAZY)
+    private List<Question> questions = new ArrayList<>();
```
> 1번
제가 ManyToOne 관계를 Answer, Question, DeleteHistory 단방향을 설정하고 나서 user에서 '각각에 대해 양방향을 설정해야 할까?' 라는 부분이 굉장히 고민이었습니다. 
생각 해 본 결과 저는 아직 연관 관계에 대해서 어떻게 양방향, 단방향을 결정지을 지 잘 모른다고 생각하게 되었습니다.

```
a 와 b 연관 관계를 파악할 때 현재는
b는 1개의 a만 필요로 한다. OneToOne
a도 1개의 b만 필요로 한다. 
```

```
b는 여러 개의 a를 필요로 한다. ManyToOne
a는 1개의 b를 필요로 한다. OneToMany 양방향
a도 여러 개의 b를 필요로 한다. ManyToMany
```

> 이런 형태로 생각을 하면서 설계를 하고 있는데요. 
a는 ~개의 b를 필요로 한다. 라는 형태로 생각을 해서 연관 관계를 결정해도 괜찮을까요?
리뷰어님이 보통 연관 관계를 설정하실 때는 어떤 기준으로 하시는 지 궁금합니다!

> 2번
제가 user에서 Answer, Question, DeleteHistory 중 어떤 것을 양방향으로 설정해야 할 지 고민하다가 reddit 모방하자고 생각이 떠올라 사이트를 참고하게 되었습니다.
사이트에서 user는 question List를 요구하고 answer는 question에 귀속되게 나오는 것 같아서 question만 양방향으로 설정했는데 괜찮은 양방향 설정이었는지 궁금합니다~!

이곳에 같이 답변을 드리겠습니다.

연관관계를 맵핑할 때 보통 다중성, 방향 등에 대해 고민을 하게 되는데요.
다중성은 동현님이 예로 들어주신 일대일, 다대일, 일대다 등이 있고 데이터베이스 기준으로 설정을 합니다.

방향성은 비즈니스 로직에서 두 객체가 서로 참조가 필요한지 여부를 고민해보면 좋을 것 같아요.
양방향인 경우 양쪽 객체 모두 신경써줘야 하기 때문에 기본적으로 단방향 맵핑을 하고 
나중에 양방향으로 객체 탐색이 필요하다고 느낄 때 추가하는 방식으로 진행을 합니다.
현재 `User`가 `Question`과 양방향으로 관계를 맺어야 하는지 고민해보시면 좋겠네요.

연관관계 맵핑 관련해서 아래 내용을 학습해보시는 것도 좋을 것 같아요 :)

[JPA 연관 관계 한방에 정리 (단방향/양방향, 연관 관계의 주인, 일대일, 다대일, 일대다, 다대다)](https://jeong-pro.tistory.com/231)

lewisseo91 replied:
감사합니다~! 연관 관계에 대해서 알쏭달쏭한 측면이 있어서 도움을 요청드렸습니다~!
글 잘 읽어보고 학습하겠습니다. 좋은 레퍼런스 주셔서 감사합니다~!!

slamdunk7575 reviewed (approved):
피드백 반영 잘해주셨네요.
질문에 대한 답변과 소소한 코멘트 남겨드렸는데 확인해보시면 좋을 것 같아요.
끝까지 화이팅 입니다~! 💪
